Handle refunded webhook event

mgascam opened the pull request:
Fixes #8085 

#### Changes proposed in this Pull Request

This pull request adds a webhook handler for charge refund events.

- Adds `process_webhook_refund` method.
- Adds `WC_Payments_Refunded_Event_Note` to handle refund-related order notes.
- Add two methods to `WC_Payments_Order_Service`: `process_order_refund` and `create_refund_for_order ` to handle the refund process.
- Refactor `process_refund` of `WC_Payment_Gateway_WCPay` to use new methods and avoid repetition.

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Refund an order in full**

1. Purchase an order
2. Perform a refund for the full amount of the order using Stripe's dashboard.
3. Access the order details page in WP Admin.
4. Check that the status of the order changed to Refunded.
5. Check that the Order note is added. Pay attention to the refund amount:
<img width="262" alt="Screenshot 2024-02-20 at 19 33 39" src="https://github.com/Automattic/woocommerce-payments/assets/1553182/fc67e061-d437-4322-89be-e7eb73e9b483">

6. Check that the Refund button is not present.


**Refund an order partially**
1. Purchase an order
2. Perform a partial refund of the order using Stripe's dashboard.
3. Access the order details page in WP Admin.
4. Check that the status of the order is not changed, should be Processing.
5. Check that the Order note is added. Pay attention to the refund amount:
<img width="261" alt="Screenshot 2024-02-20 at 19 37 33" src="https://github.com/Automattic/woocommerce-payments/assets/1553182/bb1cb954-f802-4d67-adc2-60978821a9f3">

6. Check that the Refund button is still present.

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

mgascam committed:
Add charge.refunded webhook event handler

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8200` or branch name `update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 874fbfe53340d47570bcc4a08e0fb2af22ce9330
- Build time: 2024-03-07 10:29:48 UTC

_Note: the build is updated when a new commit is pushed to this PR._

mgascam committed:
Refactor

mgascam committed:
Add unit tests

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

mgascam committed:
Create refund

mgascam committed:
Refund line items of order

mgascam committed:
Handle partial refunds (WIP)

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

mgascam committed:
Refactor in progress

mgascam committed:
Refactor

mgascam committed:
Refactor

mgascam committed:
Add tests

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

mgascam committed:
Add doc comment

mgascam committed:
Add type hints

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

mgascam committed:
Add changelog

mgascam committed:
Fix psalm error

mgascam committed:
Remove erroneous track call and minor tweaks

KarlisJ reviewed (commented):
I'm unfamiliar with the inner workings of refund in WC core but it seems strange to me that you had to create so much methods and the note. 

From my PoV, the refund execution should be the same as when triggered "normal" way. Can't we hook into existing functionality around refunds? 

Would love to see someone more familiar with WC's core refunds as reviewer here.

mgascam commented:
> From my PoV, the refund execution should be the same as when triggered "normal" way. Can't we hook into existing functionality around refunds?

Thanks @KarlisJ, I see what you mean. We have the `process_refund` method of the `WC_Payment_Gateway_WCPay` https://github.com/Automattic/woocommerce-payments/blob/e0e0afd47e0bcb6c3b5bf31efdb3dcab864c0f53/includes/class-wc-payment-gateway-wcpay.php#L2185 

This method takes care of creating the request to Stripe to process the refund, but this part is not needed to handle the event.

I'd say that there may be an opportunity to refactor the code in the `process_refund` method, to use some parts of the code that I'm adding in this pull request. Namely, the part that takes care of adding the note, setting the status of the order, and the metadata. I'll look into it, but I'm concerned that the scope of this PR will grow too much.

mgascam committed:
PR feedback

mgascam committed:
Refactor to use order service to process refund in gateway method

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

mgascam committed:
Fix using the wrong refund reference in partial refunds

mgascam committed:
Fix test

mgascam committed:
Pass instance of refund

mgascam committed:
Streamline approach

mgascam committed:
Add test for refund not found

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

mgascam committed:
Fix tests

htdat commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2281,46 +2281,13 @@ static function ( $refund ) use ( $refund_amount ) {
-				$reason,
-				$refund['id']
-			);
+		$wc_refund = WC_Payments_Utils::get_last_refund_from_order_id( $order->get_id() );
+		if ( ! $wc_refund ) {
+			// translators: %1$: order id.
+			return new WP_Error( 'wcpay_edit_order_refund_not_found', sprintf( __( 'A refund cannot be found for order: %1$s', 'woocommerce-payments' ), $order->get_id() ) );
```
❓  This is changing the original behavior when it did not throw error in case there is no last refund. 

✅  It took me a while to look at the Woo core code to confirm that `process_refund` in the Woo core abstract `WC_Payment_Gateway` will be called after `wc_create_refund`. In other words, up until this point, there is always a refund here. Otherwise, something wrong was leaked in the upstream. 

_Update_: nothing to address here, just flag this behavior change, and explain for myself why it is appropriate.

htdat commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2281,46 +2281,13 @@ static function ( $refund ) use ( $refund_amount ) {
-		$this->order_service->set_wcpay_refund_status_for_order( $order, 'successful' );
-		$order->save();
+		// If the refund was successful, add a note to the order and update the refund status.
+		$this->order_service->process_order_refund( $order, $wc_refund, $refund['id'], $refund['balance_transaction'] ?? null );
```
The method name `process_order_refund` seems misleading. The real processing of sending API requests to Stripe happened earlier? IMO, we could give it a bit long but more correct name such as `add_note_and_metadata_for_refund`?

mgascam replied:
Yes that's a fair point, followed your suggestion here https://github.com/Automattic/woocommerce-payments/pull/8200/commits/717d9202ff092bf1edfdeef55ca3c221bb1b120d thanks!

htdat commented on includes/class-wc-payments-refunded-event-note.php:
```diff
@@ -0,0 +1,109 @@
+	 *
+	 * @param float    $refunded_amount Refund amount.
+	 * @param string   $refunded_currency Refund currency.
+	 * @param string   $refund_id Refund id.
```
Should clarify this as `$wcpay_refund_id`? It's similar to the signature of  `WC_Payments_Order_Service::set_wcpay_refund_id_for_order`

mgascam replied:
I ended up removing this class 🤝

htdat commented on includes/class-wc-payments-refunded-event-note.php:
```diff
@@ -0,0 +1,109 @@
+	 *
+	 * @return string
+	 */
+	public function generate_html_note(): string {
```
It sounds like we do not need this whole new class? Instead, you can follow the same pattern `generate_payment_xxx_note` in `WC_Payments_Order_Service`?

mgascam replied:
Yes, I guess the new class does not make a great difference and adding a new method to generate notes won't hurt, addressed here https://github.com/Automattic/woocommerce-payments/pull/8200/commits/8b391f1e3ed4f5be35bc9d551facc7af98aee82c thanks!

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

htdat commented on includes/class-wc-payments-order-service.php:
```diff
@@ -1240,6 +1240,67 @@ public function capture_authorization_on_order_status_change( int $order_id ) {
+			'amount'    => $amount,
+			'reason'    => $reason,
+			'order_id'  => $order->get_id(),
+			'refund_id' => $refund_id,
```
Looking the usage of this method `create_refund_for_order`, this `$refund_id` is actually the Stripe/WCPay refund ID, which does not matter for Woo core function `wc_create_refund`. And if an `refund_id` is provided, it will try to load the refund from DB rather than to really create it - [Woo core ref](https://github.com/woocommerce/woocommerce/blob/9f7a6779f6d4134d9bc74b9655dcac6dbe108360/plugins/woocommerce/includes/wc-order-functions.php#L547-L548)

In summary, I think we would need to delete this param `$refund_id` from this method as a whole.

mgascam replied:
Great catch thanks! Addressed here https://github.com/Automattic/woocommerce-payments/pull/8200/commits/051ae5d97cc4070a1b2d4a52021700b19d1d4a30

htdat commented on includes/class-wc-payments-order-service.php:
```diff
@@ -1240,6 +1240,67 @@ public function capture_authorization_on_order_status_change( int $order_id ) {
+
+		// Set refund metadata.
+		$this->set_wcpay_refund_status_for_order( $order, 'successful' );
+		$this->set_wcpay_refund_id_for_order( $order, $refund_id );
```
The original code used param `$wc_last_refund`: 

```suggestion
		$this->set_wcpay_refund_id_for_order( $wc_refund, $refund_id );
```

--- 

I think the method name `set_wcpay_refund_id_for_order` is a bit confusing, as this needs to apply to `WC_Order_Refund` object rather than `WC_Order` object. 

Actually, the meta key `_wcpay_refund_id` for this method was originally introduced here https://github.com/Automattic/woocommerce-payments/pull/3965. And so far, the purpose of this meta key is still the same, i.e. dealing with async failed refund event. 

It would be nice if this PR can also change the name of `set_wcpay_refund_status_for_order` to `set_wcpay_refund_id_for_refund`

mgascam replied:
Oops I did not notice, it seems indeed confusing, thanks for catching it. Addressed the change of name here https://github.com/Automattic/woocommerce-payments/pull/8200/commits/ebdd3df9798513922fa767079651bfa1b0f57a25 thanks!

htdat commented on includes/class-wc-payments-webhook-processing-service.php:
```diff
@@ -783,4 +787,66 @@ private function get_failure_message_from_error( $error ):string {
+		// This is to avoid processing the same refund multiple times in case of a webhook retry.
+		if ( $refund_id === $this->order_service->get_wcpay_refund_id_for_order( $order ) ) {
+			return;
+		}
```
I think the correct logic here would be similar this part in method [process_webhook_refund_updated](https://github.com/Automattic/woocommerce-payments/blob/6b04ff459bbaffb532d04b95b1beea3c73817c0d/includes/class-wc-payments-webhook-processing-service.php#L301-L316). I.e: check if the refund ID from the incoming webhook event matches with any existing refund objects (which are attached to this order), and just continue processing if there is no match.

mgascam replied:
Indeed, that's a more precise approach, followed your suggestion here https://github.com/Automattic/woocommerce-payments/pull/8200/commits/8d29dba8f1c5cb49f463f1074a6968838ad21469 thanks!

htdat reviewed (commented):
Great work to figure out the solution for this PR. I've left a few comments, some of them are critical to address.

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

mgascam committed:
PR feedback: remove unneeded parameter

mgascam committed:
PR feedback: discard WC_Payments_Refunded_Event_Note class

mgascam committed:
PR feedback: refactor method name

mgascam committed:
PR feedback

mgascam committed:
PR feedback: refactor method name

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

mgascam commented:
> Great work to figure out the solution for this PR. I've left a few comments, some of them are critical to address.

Thanks for your thorough review @htdat it helped me a lot. I hope I addressed your comments, ready for a new round of reviews.

mgascam committed:
Amend doc comment

htdat commented on includes/class-wc-payments-order-service.php:
```diff
@@ -1575,6 +1634,54 @@ private function generate_dispute_closed_note( $charge_id, $status ) {
+	 *
+	 * @param float    $refunded_amount Amount refunded.
+	 * @param string   $refunded_currency Refund currency.
+	 * @param string   $refund_id Refund ID.
```
```suggestion
	 * @param string   $wcpay_refund_id WCPay WCPay refund ID.
```

mgascam replied:
Done https://github.com/Automattic/woocommerce-payments/pull/8200/commits/c06b289ef949ef8a1ae1fda6704ba2a830e7fb75 thanks!

htdat commented on includes/class-wc-payments-order-service.php:
```diff
@@ -687,7 +687,7 @@ public function get_wcpay_intent_currency_for_order( $order ) : string {
 	 * @throws Order_Not_Found_Exception
 	 */
-	public function set_wcpay_refund_id_for_order( $order, $wcpay_refund_id ) {
+	public function set_wcpay_refund_id_for_refund( $order, $wcpay_refund_id ) {
```
```suggestion
	public function set_wcpay_refund_id_for_refund( $wc_refund, $wcpay_refund_id ) {
```

Also update PHPdoc (just my suggestions but use your own judgemnt :D )

- Type `WC_Order_Refund` for the first param `$wc_refund`
- Description of method: Set WCPay refund ID as metedata for refund object.

mgascam replied:
I admit I was a bit confused by the fact that we call `$this->get_order( $order )` in the body of this method. But then I understood the reason behind is to be able to use mocks for the sake of testing. I decided to change the references to `$order` in the body of the method to `$wc_refund`, hope it makes sense as well. You can check it in this commit https://github.com/Automattic/woocommerce-payments/pull/8200/commits/bc135bca5f8519a651d627bfe035cc613bc4de36 thanks!

htdat commented on includes/class-wc-payments-webhook-processing-service.php:
```diff
@@ -783,4 +787,71 @@ private function get_failure_message_from_error( $error ):string {
+				'order_not_found'
+			);
+		}
+		// This is to avoid processing the same refund multiple times in case of a webhook retry.
```
More precisely: 
```suggestion
		// Only care about refunds that are triggered externally, i.e. outside WP Admin.
```

IMO, we would also change this method name to `process_webhook_refund_trigged_externally`. Though it's pretty long but it conveys the full purpose of this PR as well as this method.

mgascam replied:
Thanks for the suggestion, I agree that it's more clear this way. I also expanded the comments to point to the place where WP Admin refunds are handled. See [9224dc1](https://github.com/Automattic/woocommerce-payments/pull/8200/commits/9224dc1f71e12ad1e908f9545c7b5e7f58459984).

htdat commented on changelog/update-8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: add
+
+Adds a handler for the refunded webhook event
```
Nitpick: 
```suggestion
Handle refunds triggered externally, outside WP Admin
```

mgascam replied:
👍 https://github.com/Automattic/woocommerce-payments/pull/8200/commits/8424d7336d64bc76018c958f4e67d6bf816a4383

htdat reviewed (commented):
Thanks for addressing the previous feedback. I think it's very close for the merge. 

✅ Manual tests work well in general.
❓ However, the format of amount and currency in the refund note are different when refunds are triggering in WP Admin vs externally (Stripe Dashboard) - 
[see this screenshot](https://github.com/Automattic/woocommerce-payments/assets/10045087/53803bde-c1c1-4522-a7db-fd65ece6a6db). Could you have a look at them? 
✅ Unit tests. 
❓ Have some other nitpick comments.

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

mgascam committed:
Fix amount formatting discrepancy

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

mgascam committed:
PR feedback: refactor method params

mgascam committed:
PR feedback: refactor param name

mgascam committed:
PR feedback: refine comment

mgascam committed:
PR feedback: refactor method name

mgascam committed:
PR feedback: amend changelog

mgascam commented:
> ❓ Have some other nitpick comments.

Thanks again @htdat I believe I addressed all your comments, and marked them as resolved. 

> ❓ However, the format of amount and currency in the refund note are different when refunds are triggering in WP Admin vs externally (Stripe Dashboard)

I looked into this one, but could not reproduce it on my side. The format of the refunded amount looks the same in the order notes, regardless of how the refund was triggered 🤔. 

Looking at the two flows, I found a difference ([ref](https://github.com/Automattic/woocommerce.com/blob/93a5f9483f3ac340681f0c1662d0733933e828c6/plugins/woocommerce/includes/class-wc-ajax.php#L2147)) in how the refund amount is formatted when the `WC_Refund` is created so I'm doing the same in this commit https://github.com/Automattic/woocommerce-payments/pull/8200/commits/122ea128e7c3958788276548a4a10675a7e3a3ed. Can you please take a look on your end and let me know if this makes any difference?

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes

mgascam committed:
Fix linter issue

htdat commented:
> Looking at the two flows, I found a difference ([ref](https://github.com/Automattic/woocommerce.com/blob/93a5f9483f3ac340681f0c1662d0733933e828c6/plugins/woocommerce/includes/class-wc-ajax.php#L2147)) in how the refund amount is formatted when the WC_Refund is created so I'm doing the same in this commit https://github.com/Automattic/woocommerce-payments/commit/122ea128e7c3958788276548a4a10675a7e3a3ed. Can you please take a look on your end and let me know if this makes any difference?

This does not create much difference as eventually, the `generate_payment_refunded_note` method will use the same arguments. It turns out that this difference comes from two elements: 
- some configurations on my site is set up with `Currency position: Right`, and `Number of decimals: 0` under wp-admin/admin.php?page=wc-settings 
- somehow `wc_price` returns different formats when running in AJAX request (WP Admin request) and a REST request (incoming webhook event). 

In other words, this issue is not from your PR, but it would be nice if you can create an issue to log it.

htdat reviewed (approved):
Pre-approve PR! You can merge this PR after all CI runs pass.

mgascam committed:
Merge branch 'develop' into update/8085-refunds-revisit-events-handling-externally-sourced-refunds-are-not-changing-order-state-not-rtacked-in-order-notes
